pkg/icingadb/schema.go

```diff
@@ -48,11 +69,85 @@ func CheckSchema(ctx context.Context, db *database.DB) error {
 		// Since these error messages are trivial and mostly caused by users, we don't need
 		// to print a stack trace here. However, since errors.Errorf() does this automatically,
 		// we need to use fmt instead.
-		return fmt.Errorf(
-			"unexpected database schema version: v%d (expected v%d), please make sure you have applied all database"+
-				" migrations after upgrading Icinga DB", version, expectedDbSchemaVersion,
+		return fmt.Errorf("%w: v%d (expected v%d), please make sure you have applied all database"+
+			" migrations after upgrading Icinga DB", ErrSchemaMismatch, version, expectedDbSchemaVersion,
 		)
 	}
 
 	return nil
 }
+
+// ImportSchema performs an initial schema import in the db.
+//
+// This function assumes that no schema exists. So it should only be called after a prior CheckSchema call.
+func ImportSchema(
+	ctx context.Context,
+	db *database.DB,
+	databaseSchemaDir string,
+) error {
+	var schemaFileDirPart string
+	switch db.DriverName() {
+	case database.MySQL:
+		schemaFileDirPart = "mysql"
+	case database.PostgreSQL:
+		schemaFileDirPart = "pgsql"
+	default:
+		return fmt.Errorf("unsupported database driver %q", db.DriverName())
+	}
+
+	tx, err := db.BeginTxx(ctx, nil)
+	if err != nil {
+		return errors.Wrap(err, "can't start database transaction")
+	}
+	defer func() { _ = tx.Rollback() }()
+
+	schemaFile := path.Join(databaseSchemaDir, schemaFileDirPart, "schema.sql")
+	schema, err := os.ReadFile(schemaFile) // #nosec G304 -- path is constructed from "trusted" command line user input
+	if err != nil {
+		return errors.Wrapf(err, "can't open schema file %q", schema)
+	}
+
+	switch db.DriverName() {
+	case database.MySQL:
+		for _, query := range database.MysqlSplitStatements(string(schema)) {
+			err := retry.WithBackoff(
+				ctx,
+				func(ctx context.Context) error {
+					if _, err := tx.ExecContext(ctx, query); err != nil {
+						return database.CantPerformQuery(err, query)
+					}
+					return nil
+				},
+				retry.Retryable,
+				backoff.NewExponentialWithJitter(128*time.Millisecond, 1*time.Minute),
+				db.GetDefaultRetrySettings())
+			if err != nil {
+				return errors.Wrap(err, "can't perform schema import")
+			}
+		}
+
+	case database.PostgreSQL:
+		err := retry.WithBackoff(
+			ctx,
+			func(ctx context.Context) error {
+				if _, err := tx.ExecContext(ctx, string(schema)); err != nil {
+					return err
+				}
+				return nil
+			},
+			retry.Retryable,
+			backoff.NewExponentialWithJitter(128*time.Millisecond, 1*time.Minute),
+			db.GetDefaultRetrySettings())
+		if err != nil {
+			return errors.Wrap(err, "can't perform schema import")
+		}
+
```

Comment:
All other switch statements introduced have no line break between the cases. I opt to remove them here as well.